code/modules/antagonists/mind_flayer/powers/flayer_passives.dm

```diff
@@ -0,0 +1,307 @@
+	var/power_type = FLAYER_UNOBTAINABLE_POWER
+	///The base cost of an ability, used to calculate how much upgrades should cost.
+	var/base_cost = 30
+	///How much it will cost to upgrade this passive.
+	var/current_cost
+	///If the passive is for a specific class, or CATEGORY_GENERAL if not
+	var/category = CATEGORY_GENERAL
+	///If the passive requires prerequisites, currently only important for badass.
+	var/stage = 0
+	///A brief description of what the ability's upgrades do
+	var/upgrade_info = "TODO add upgrade text for this passive"
+	/// Does this passive need to process
+	var/should_process = TRUE
+
+/datum/mindflayer_passive/New()
+	. = ..()
+	current_cost = base_cost
+	if(should_process)
+		START_PROCESSING(SSobj, src)
+
+/datum/mindflayer_passive/Destroy(force, ...)
+	. = ..()
+	if(!flayer)
+		return
+	on_remove()
+	STOP_PROCESSING(SSobj, src)
+
+///This is where most passive's effects get applied
+/datum/mindflayer_passive/proc/on_apply()
+	SHOULD_CALL_PARENT(TRUE)
+	flayer.send_swarm_message(level ? upgrade_text : gain_text) //This will only be false when level = 0, when first bought
+	level++
+	current_cost = base_cost * (level + 1)
+	return TRUE
+
+/datum/mindflayer_passive/proc/on_remove()
+	return
+
+/datum/mindflayer_passive/process()
+	return
+
+//SELF-BUFF PASSIVES
+/datum/mindflayer_passive/armored_plating
+	name = "Armored Plating"
+	purchase_text = "Increases our natural armor."
+	upgrade_text = "The swarm adds more layers of armored nanites, strengthening the plating even more."
+	upgrade_info = "Further increases base armor by 10"
+	gain_text = "You feel your chassis being reinforced by the swarm."
+	power_type = FLAYER_PURCHASABLE_POWER
+	max_level = 3
+	var/armor_value = 0
+
+/datum/mindflayer_passive/armored_plating/on_apply()
+	..()
+	var/owner_armor = owner.dna.species.armor
+	var/temp_armor_value = owner_armor - (5 * (level - 1)) // We store our current armor value here just in case they already have armor
+	armor_value = temp_armor_value + 5 * level
+	owner.dna.species.armor = armor_value
+
+/datum/mindflayer_passive/armored_plating/on_remove()
+	owner.dna.species.armor -= armor_value
+
+/datum/mindflayer_passive/fluid_feet
+	name = "Fluid Feet"
+	purchase_text = "Mute your footsteps, then upgrade to become mostly unslippable."
+	upgrade_text = "Your feet become even more malleable, seemingly melting into the floor; you feel oddly stable."
+	gain_text = "Your limbs start slowly melting into the floor."
+	upgrade_info = "Become nearly unslippable."
+	power_type = FLAYER_PURCHASABLE_POWER
+	max_level = 2
+
+/datum/mindflayer_passive/fluid_feet/on_apply()
+	..()
+	switch(level)
+		if(POWER_LEVEL_ONE)
+			qdel(owner.GetComponent(/datum/component/footstep))
+		if(POWER_LEVEL_TWO)
+			ADD_TRAIT(owner, TRAIT_NOSLIP, UNIQUE_TRAIT_SOURCE(src))
+
+/datum/mindflayer_passive/fluid_feet/on_remove()
+	owner.AddComponent(/datum/component/footstep, FOOTSTEP_MOB_HUMAN, 1, -6)
```

Comment:
Can you make sure it's exactly the same as before, and is there a concern about it being added twice?

Reply:
Humans can only ever have FOOTSTEP_MOB_HUMAN, and on_remove can't be called multiple times for the same passive.